lib/querystring.js

```diff
@@ -319,26 +320,36 @@ function parse(qs, sep, eq, options) {
           key = decodeStr(key, decode);
         if (valEncoded)
           value = decodeStr(value, decode);
-        // Use a key array lookup instead of using hasOwnProperty(), which is
-        // slower
-        if (keys.indexOf(key) === -1) {
-          obj[key] = value;
-          keys[keys.length] = key;
-        } else {
-          const curValue = obj[key];
-          // A simple Array-specific property check is enough here to
-          // distinguish from a string value and is faster and still safe since
-          // we are generating all of the values being assigned.
-          if (curValue.pop)
-            curValue[curValue.length] = value;
-          else
-            obj[key] = [curValue, value];
+
+        if (lastPos === 0 || lastPos - posIdx > sepLen) {
+          // Use a key array lookup instead of using hasOwnProperty(), which is
+          // slower
+          if (keys.indexOf(key) === -1) {
+            obj[key] = value;
+            keys[keys.length] = key;
+          } else if (lastPos - sepIdx > sepLen) {
+            const curValue = obj[key];
+            if (curValue) {
+              // A simple Array-specific property check is enough here to
+              // distinguish from a string value and is faster and still safe
+              // since we are generating all of the values being assigned.
+              if (curValue.pop)
+                curValue[curValue.length] = value;
+              else
+                obj[key] = [curValue, value];
+            }
+          }
+        } else if (posIdx === 0) {
+          // A pair with repeated sep could be added into obj in the first loop
+          // and it should be deleted
+          delete obj[key];
```

Comment:
Note `benchmark/url-parse.js` can't really capture the performance hit because they are all making monomorphic calls, which would be rare in real word usage.